riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java

```diff
@@ -0,0 +1,183 @@
+        };
+    }
+
+    private static ThrowingConsumer<ThrowableProblem> propagate() {
+        return entity -> {
+            throw entity;
+        };
+    }
+
+    static class GatewayException extends RuntimeException {
+
+        public GatewayException(Throwable cause) {
+            super(cause);
+        }
+    }
+
+    private final Rest rest;
+
+    @Autowired
+    public NakadiGateway(final Rest rest) {
+        this.rest = rest;
+    }
+
+    public void stream(final String owningApplication, final List<String> eventTypes, final EventConsumer consumer)
+            throws InterruptedException, ExecutionException, IOException {
+        stream(subscribe(new Subscription(null, owningApplication, eventTypes)), consumer);
+    }
+
+    private void stream(final Subscription subscription, final EventConsumer consumer)
+            throws InterruptedException, ExecutionException, IOException {
+        rest.get("/subscriptions/{id}/events", subscription.getId())
+                .dispatch(series(),
+                        on(SUCCESSFUL).call(responseEntityOf(streamOf(Batch.class)),
+                                process(subscription, consumer)),
+                        ON_FAILURE_BINDING)
+                .get();
+    }
+
+    private Subscription subscribe(final Subscription subscription)
+            throws InterruptedException, ExecutionException, IOException {
+        final Capture<Subscription> capture = Capture.empty();
+        rest.post("/subscriptions")
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(subscription)
+                .dispatch(series(),
+                        on(SUCCESSFUL).call(Subscription.class, capture),
+                        ON_FAILURE_BINDING)
+                .get();
+        return capture.retrieve();
+    }
+
+    private ThrowingConsumer<ResponseEntity<Stream<Batch>>> process(final Subscription subscription,
+            final EventConsumer consumer) {
+        return entity -> {
+            @SuppressWarnings("resource")
+            final Stream<Batch> stream = entity.getBody();
+            try {
+                final HttpHeaders headers = entity.getHeaders();
+                stream.map(batch -> consume(consumer, batch))
+                        .filter(cursor -> cursor != null)
+                        .forEach(cursor -> commit(subscription, headers, cursor));
+            } finally {
+                stream.close();
+            }
+        };
+    }
+
+    private Cursor consume(final EventConsumer consumer, final Batch batch) {
+        Cursor cursor = batch.getCursor();
+        if (cursor == null) {
+            LOG.debug("invalid batch: [{}]", batch);
+            return cursor;
+        }
+        List<Event> events = batch.getEvents();
+        if (events == null || events.isEmpty()) {
+            LOG.debug("empty batch: [{}]", batch);
+            return cursor;
```

Comment:
I think returning null would be cleaner here, since there is nothing to commit, is there? @AlexanderYastrebov Thoughts, since you talked to the Nakadi guys about this?